### ISSUE TITLE
AutoFormat.RemoveEmpty strips video and audio elements

### ISSUE THREAD

**Issue by janboddez** [xemlock/htmlpurifier-html5#64]
Because of the allowed elements inside figure? Do they not yet include HTML5 elements?

**Comment by xemlock** [comment]
Please provide a minimal working example that would show the issue, as I'm not able to replicate this.

`<figure>` [accepts Flow elements as its children](https://github.com/xemlock/htmlpurifier-html5/blob/master/library/HTMLPurifier/ChildDef/HTML5/Figure.php#L25), and `<video>` is [defined as a Flow element](https://github.com/xemlock/htmlpurifier-html5/blob/master/library/HTMLPurifier/HTMLModule/HTML5/Media.php#L24).

**Comment by janboddez** [comment]
You're right, turns out it's due to my use of `$config->set('AutoFormat.RemoveEmpty', true)`. First time I noticed this, which is why I wrongly blamed `figure`.

I still didn't expect that to affect `video`, though. It doesn't remove (self-closing) `img`s ... Or (equally self-closing) `audio`s. In `video`'s case, it doesn't seem to matter. Both `<video src="something"></video>` and `<video src="something" />` seem to get removed. (Not sure about one or more `source` tags wrapped in `video`. I _think_ that would be okay.)

Now, I'm not sure if there's a way to prevent this? Other than manually override, on my side, I mean, `AutoFormat.RemoveEmpty.Predicate`.

Quick update, this seems to work, for now:
```
$config->set('AutoFormat.RemoveEmpty.Predicate', [
    'colgroup' => [],
    'th'       => [],
    'td'       => [],
    'iframe'   => [0 => 'src'],
    'video'    => [], // Bit of a hack, no?
]);
```

**Comment by xemlock** [comment]
`<img>` and `<video>`/`<audio>` tags are structurally different. `<img>` is a self-closing tag, whereas the other two are not. RemoveEmpty strategy doesn't apply on self-closing tags, this is why `img` is not affected by it (and neither is `<source>`).

By default RemoveEmpty strategy has no predefined rules for `<video>` and `<audio>`, so yeah, the only reasonable solution (and definitely not a hack) is to include them in the `AutoFormat.RemoveEmpty.Predicate` setting, which is exactly what you proposed, i.e.:

```php
$config->set('AutoFormat.RemoveEmpty.Predicate', [
    'colgroup' => [],
    'th'       => [],
    'td'       => [],
    'iframe'   => ['src'],
    'video'    => [],
    'audio'    => [],
]);
```

BTW, I've see no difference between how `<video>` and `<audio>` elements are handled. Both are automatically removed if `AutoFormat.RemoveEmpty` is set to `true`. But it may depend on the chosen lexer or the specific version of php-dom you're using.

**Comment by xemlock** [comment]
Resolved by #65.
